File: com/white/SwapNodesInPairs.java
package com.white;

public class SwapNodesInPairs {

    // 24. Swap Nodes in Pairs
    // To clarify, if there's odd number of nodes,
    // the left-out one would not be swapped with any other node.
    
    // There should be better ways to write this
    public static ListNode swapPairs(ListNode head) {
        
        if (head==null) {
            return head;
        }
        ListNode dummyHead = new ListNode(0);
        ListNode curr = head;
        ListNode prev = dummyHead;
        ListNode next = curr.next;
        prev.next = curr;
        int count = 1;

        while (next!=null) {
            if (count==0) {
                prev = prev.next;
                curr = curr.next;
                next = next.next;
                count++;
            } else {
                prev.next = next;
                curr.next = next.next;
                next.next = curr;

                prev = prev.next;
                next = curr.next;
                count--;
            }
        }

        return dummyHead.next;
    }
}

File: LeetcodeCore/SwapNodesInPairs.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class SwapNodesInPairs
    {
        // 24. Swap Nodes in Pairs
        public ListNode SwapPairs(ListNode head)
        {

            var dummy = new ListNode(0, head);
            RecursiveHelper(dummy, head);
            return dummy.next;
        }

        private void RecursiveHelper(ListNode pre, ListNode curr)
        {
            if (curr == null || curr.next == null)
                return;

            pre.next = curr.next;
            curr.next = pre.next.next;
            pre.next.next = curr;

            RecursiveHelper(curr, curr.next);
        }
    }
}
